### ISSUE TITLE
Black screen when transition from map to map

### ISSUE THREAD

**Issue by 0x4E69676874466F78** [w23/xash3d-fwgs#250]
![изображение](https://user-images.githubusercontent.com/4449851/144505602-6e8676fc-19d9-4ee3-bb7a-288972b210ce.png)
Нужно сохранять последний кадр перед перемещением.

**Comment by w23** [comment]
(почему у меня метки девственный `visual bug` вместо чад `МИНУС ГЛОЗА`?!)

**Comment by 0x4E69676874466F78** [comment]
@w23 ну вообще семантика разная, не везде визуалбаг это минус глаза, но тут могу добавить.

**Comment by zgdump** [comment]
Поковырял проблему. Судя по всему, баг не в рендере, а в самом движке.

При смене уровня через цепочку вызовов и команд происходит вызов [`SCR_MakeLevelShot`](https://github.com/w23/xash3d-fwgs/blob/c5bbcae24258041f17a332e0bb2cd95bdb123b4c/engine/client/cl_scrn.c#L239), где происходит проверка поля `scrshot_request`. И по какому-то недоразумению, это поле никто не обновляет перед вызовом функции и команда `levelshot` не вызывается.

Для проверки я выпилил проверку `scrshot_request` в этой функции и в обработчике команды, чтобы скриншот уровня можно было сделать через консоль. И скриншот создался!

Скриншот уровня создался и в gl, и в vk. Под капотом все скриншоты делаются ровно через одну функцию `VID_ScreenShot` в рендере.

Как доказательство, прикладываю скриншот уровня из `valve/levelshots/c1a0_16x9.bmp`, конвертированный в Jpeg:

![c1a0_16x9](https://github.com/w23/xash3d-fwgs/assets/60278551/36890ca1-b281-4fb5-91bf-7fd912787793)

И теперь, если заходить на уровень, всегда будет отрисовываться именно эта картинка.

Как я понимаю, все ишью с названием [чёрный экран при смене уровня](https://github.com/FWGS/xash3d-fwgs/issues?q=is%3Aissue+is%3Aopen+%D1%87%D1%91%D1%80%D0%BD%D1%8B%D0%B9+) связанны именно с этим.

**Comment by zgdump** [comment]
Дополню. Эту проблему стоит или переадресовать в двигло, позакрывав все связанные ишью. Или починить и бэкпортировать, это не должно быть сложным. Единственное, могут всплыть ещё какие-то проблемы. Например, меня лично смущает, что скриншот не удаляется после смены уровня...

**Comment by 0x4E69676874466F78** [comment]
Как я понял из обсуждения levelshots для другого. В случае с vulkan проблема в том что не делается захват последнего кадра. В GL как я понимаю там попросту очистки кадра не происходит, поэтому последний кадр игры отображается при загрузке. В случае с VK идёт очистика кадра происходит всегда поэтому чёрный или красный экран видим.

**Comment by zgdump** [comment]
@0x4E69676874466F78 да, levelshots оказались для другого. Зря побеспокоил людей. Но всё и правда выглядело так, будто они нужны для переходов между уровнями... Ну и ладно.

Но я всё ещё считаю, что это бага движка, а не рендера: нельзя полагаться на то, что рендер не отчистит экран и сам как-то догадается оставить предыдущий кадр при загрузке уровня. В ref gl, как видно, это работает через раз: у меня на Alder Lake GT-1 не работает, на Vega 7 работает, на телефонах по разному. Напрашивается какое-то другое решение, более явное.

**Comment by 0x4E69676874466F78** [comment]
@zgdump если это не работает где-то на GL значит есть бага в GL и надо там завести задачку на эту тему.
Рендеру по идее довольно просто понять когда нужно не стирать кадр (или захватить его для вывода) при смене карты. Есть вызов R_NewMap где по идее можно сделать захват кадра.

**Comment by w23** [comment]
Не очищать предыдущий кадр это условно дорогая операция. Не уверен, что она в вулкане вообще доступна, кроме как руками копировать из предыдущего изображения. Что тоже не так просто -- его надо создавать тогда с usage битиками правильными, и они желанные могут быть incompatible (мы там для скриншотов чем-то подобным страдали)

**Comment by 0x4E69676874466F78** [comment]
@w23 да помню 400 строк на вулкане вместо 1 строки на GL. 
Если я правильно догадываюсь о механике, то абстракции "не очищать" просто нет, зато есть буфер с двумя кадрами, который переключается по мере готовности того или иного кадра, таким образом когда мы ждём переключение на следующий кадр там в лучшем случае предпредыщущий кадр, а в худшем мусор.
Следовательно надо каким-то образом копировать последний активный кадр в ожидаемый следующий.
Правильно понимаю?

**Comment by w23** [comment]
Посмотрели на это на стриме E323. Можно:
- вынести всю рендерню из `vk_framectl.c` `enqueueRendering()` поближе к `vk_render.c` (?)
- и там внутри, зная о том, есть ли у нас что порисовать, принимать решение о копировании из предыдущего кадра.
- Альтернативно: ничего никуда не переносить особо, только вынести знание о пустоте сцены.
- Нужно правильно следить за изображением предыдущего кадра, иметь к нему доступ, и уметь с ним синхронизоваться.
